### ISSUE TITLE
StringIO module truncate behavior of current position

### ISSUE THREAD

**Issue by bf28a09e-6d72-4c82-a4b1-8cc669e32358** [python/cpython#74436]
BPO | [30250](https://bugs.python.org/issue30250)
--- | :---
Nosy | @albertz, @MojoVampire, @ZackerySpytz

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2020-05-08.23:00:51.294>
created_at = <Date 2017-05-03.09:21:26.903>
labels = ['expert-IO', 'docs']
title = 'StringIO module truncate behavior of current position'
updated_at = <Date 2020-05-08.23:00:51.293>
user = 'https://github.com/albertz'
```

bugs.python.org fields:
```python
activity = <Date 2020-05-08.23:00:51.293>
actor = 'zach.ware'
assignee = 'docs@python'
closed = True
closed_date = <Date 2020-05-08.23:00:51.294>
closer = 'zach.ware'
components = ['Documentation', 'IO']
creation = <Date 2017-05-03.09:21:26.903>
creator = 'Albert.Zeyer'
dependencies = []
files = []
hgrepos = []
issue_num = 30250
keywords = []
message_count = 3.0
messages = ['292866', '292933', '368479']
nosy_count = 4.0
nosy_names = ['docs@python', 'Albert.Zeyer', 'josh.r', 'ZackerySpytz']
pr_nums = []
priority = 'normal'
resolution = 'out of date'
stage = 'resolved'
status = 'closed'
superseder = None
type = None
url = 'https://bugs.python.org/issue30250'
versions = ['Python 2.7']
```

</p></details>

**Comment by bf28a09e-6d72-4c82-a4b1-8cc669e32358** [comment]
The doc says that StringIO.truncate should not change the current position.
Consider this code:

  try:
    import StringIO
  except ImportError:
    import io as StringIO
  buf = StringIO.StringIO()
  assert_equal(buf.getvalue(), "")
  print("buf: %r" % buf.getvalue())

```py
  buf.write("hello")
  print("buf: %r" % buf.getvalue())
  assert_equal(buf.getvalue(), "hello")
  buf.truncate(0)
  print("buf: %r" % buf.getvalue())
  assert_equal(buf.getvalue(), "")

  buf.write("hello")
  print("buf: %r" % buf.getvalue())
  assert_equal(buf.getvalue(), "\x00\x00\x00\x00\x00hello")
  buf.truncate(0)
  print("buf: %r" % buf.getvalue())
  assert_equal(buf.getvalue(), "")
```



On Python 3.6, I get the output:

buf: ''
buf: 'hello'
buf: ''
buf: '\\x00\\x00\\x00\\x00\\x00hello'

On Python 2.7, I get the output:

buf: ''
buf: 'hello'
buf: ''
buf: 'hello'


Thus it seems that Python 2.7 StringIO.truncate does actually resets the position for this case or there is some other bug in Python 2.7. At least from the doc, it seems that the Python 3.6 behavior is the expected behavior.

**Comment by 99ffcaa5-b43b-4e8e-a35e-9c890007b9cd** [comment]
For the record, the StringIO module is Python 2 only (replaced by io.StringIO and io.BytesIO for Py3, which are also available on 2.6 and higher if you need consistency), and only documents truncate by reference to the generic docs for the built-in file object. Admittedly, the built-in file object docs say the position shouldn't change on truncate, but given:

1. StringIO is deprecated and no longer maintained
2. Code using it may rely on this incorrect behavior

I can't see fixing this beyond possibly mentioning the quirk in the 2.x StringIO docs. Since io.StringIO is correct, there is nothing to fix in 3.x.
